File: nim/sim.nim
import op

static:
  assert int(OpCode.COUNT) == 7

proc simulateProgram*(program: openArray[Op]) =
  var stack: seq[Word] = @[]
  var i: Word = 0
  while i < Word(program.len):
    let op = program[i]
    case op.code:
    of OpCode.PUSH:
      stack.add(op.operand)
      i += 1
    of OpCode.PLUS:
      let b = stack.pop()
      let a = stack.pop()
      stack.add(a + b)
      i += 1
    of OpCode.MINUS:
      let b = stack.pop()
      let a = stack.pop()
      stack.add(a - b)
      i += 1
    of OpCode.DUMP:
      let val = stack.pop()
      echo val
      i += 1
    of OpCode.EQ:
      let b = stack.pop()
      let a = stack.pop()
      stack.add(Word(a == b))
      i += 1
    of OpCode.IF:
      let a = stack.pop()
      if a == 0:
        i = op.operand
      else:
        i += 1
    of OpCode.END:
      i += 1
    else:
      raiseAssert("unreachable")

File: rust/morth/src/sim.rs
use std::io::{Read, Write};

use crate::{
    lex::lex_program,
    op::{Op, OpCode},
    parse::parse_token_as_op,
    Error, Result, Value,
};

pub fn simulate_program(
    source_filename: &str,
    source: &mut dyn Read,
    output: &mut dyn Write,
) -> Result<()> {
    let program = lex_program(source_filename, source)?
        .into_iter()
        .map(parse_token_as_op)
        .collect::<Result<Vec<_>>>()?;
    simulate(&program, output)?;
    Ok(())
}

fn simulate(program: &[Op], output: &mut dyn Write) -> Result<()> {
    let mut stack = Vec::<Value>::new();
    let mut ip = 0;
    while ip < program.len() {
        match program[ip].code {
            OpCode::Push(value) => {
                stack.push(value);
                ip += 1;
            }
            OpCode::Plus => {
                let b = stack
                    .pop()
                    .ok_or_else(|| Error::StackUnderflow(program[ip].location.clone()))?;
                let a = stack
                    .pop()
                    .ok_or_else(|| Error::StackUnderflow(program[ip].location.clone()))?;
                stack.push(a + b);
                ip += 1;
            }
            OpCode::Minus => {
                let b = stack
                    .pop()
                    .ok_or_else(|| Error::StackUnderflow(program[ip].location.clone()))?;
                let a = stack
                    .pop()
                    .ok_or_else(|| Error::StackUnderflow(program[ip].location.clone()))?;
                stack.push(a - b);
                ip += 1;
            }
            OpCode::Dump => {
                let value = stack
                    .pop()
                    .ok_or_else(|| Error::StackUnderflow(program[ip].location.clone()))?;
                writeln!(output, "{}", value)?;
                ip += 1;
            }
        }
    }
    Ok(())
}
